File: app/src/main/java/com/summer/itis/summerproject/ui/cards/add_card/AddCardActivity.java
package com.summer.itis.summerproject.ui.cards.add_card;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.SeekBar;
import android.widget.TextView;

import com.arellomobile.mvp.presenter.InjectPresenter;
import com.summer.itis.summerproject.R;
import com.summer.itis.summerproject.api.Card;
import com.summer.itis.summerproject.model.pojo.opensearch.Item;
import com.summer.itis.summerproject.model.pojo.query.Page;
import com.summer.itis.summerproject.ui.base.BaseActivity;
import com.summer.itis.summerproject.ui.tests.add_test.AddTestActivity;
import com.summer.itis.summerproject.utils.Const;

import java.util.List;

import static com.summer.itis.summerproject.utils.Const.gsonConverter;

public class AddCardActivity extends BaseActivity implements AddCardView,View.OnClickListener {

    public final static String CARD_EXTRA = "card";

    private Card card;

    private TextView tvPrestige;
    private TextView tvHp;
    private TextView tvSupport;
    private TextView tvStrength;
    private TextView tvIntelligence;
    private SeekBar sbPrestige;
    private SeekBar sbSupport;
    private SeekBar sbHp;
    private SeekBar sbStrength;
    private SeekBar sbIntelligence;
    private Toolbar toolbar;

    private String wikiUrl;
    private String photoUrl;
    private String name;

    @InjectPresenter
    AddCardPresenter presenter;

    private Item item;

    public static String ITEM_JSON = "item_json";

    public static void start(@NonNull Activity activity, Item item) {
        Intent intent = new Intent(activity, AddTestActivity.class);
        String itemJson = gsonConverter.toJson(item);
        intent.putExtra(ITEM_JSON,itemJson);
        activity.startActivity(intent);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_add_card);

        card = new Card();
        item = gsonConverter.fromJson(getIntent().getStringExtra(ITEM_JSON),Item.class);
        card.setWikiUrl(item.getUrl().getContent());
        card.setDescription(item.getDescription().getContent());
        presenter = new AddCardPresenter(this);
        presenter.query(item.getText().getContent());
        initViews();
        setListeners();
    }

    private void initViews() {
        findViews();
//        supportActionBar(toolbar);
        setSupportActionBar(toolbar);
        setBackArrow(toolbar);

    }

    private void setListeners() {
        sbStrength.setOnClickListener(this);
        sbSupport.setOnClickListener(this);
        sbPrestige.setOnClickListener(this);
        sbIntelligence.setOnClickListener(this);
        sbHp.setOnClickListener(this);

    }

    @Override
    public void setQueryResults(List<Page> list) {
        Page page = list.get(0);
        card.setName(page.getTitle());
        card.setPhotoUrl(page.getOriginal().getSource());
        card.setExtract(page.getExtract().getContent());

    }

    @Override
    public void handleError(Throwable throwable) {

    }

    private void findViews() {
        toolbar = findViewById(R.id.toolbar);
        tvHp = findViewById(R.id.tvHp);
        tvIntelligence = findViewById(R.id.tvIntelligence);
        tvPrestige = findViewById(R.id.tvPrestige);
        tvSupport = findViewById(R.id.tvSupport);
        tvStrength = findViewById(R.id.tvStrength);
        sbHp = findViewById(R.id.seekBarHp);
        sbIntelligence = findViewById(R.id.seekBarIntelligence);
        sbPrestige = findViewById(R.id.seekBarPrestige);
        sbSupport = findViewById(R.id.seekBarSupport);
        sbStrength = findViewById(R.id.seekBarStrength);
    }

    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.edit_menu, menu);

        MenuItem checkItem = menu.findItem(R.id.action_check);
        checkItem.setOnMenuItemClickListener(new MenuItem.OnMenuItemClickListener() {
            @Override
            public boolean onMenuItemClick(MenuItem item) {
                Intent intent = new Intent();
                String cardJson = gsonConverter.toJson(card);
                intent.putExtra(CARD_EXTRA, cardJson);
                setResult(RESULT_OK, intent);
                finish();

                return true;
            }
        });

        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public void onClick(View view) {
        SeekBar seekBar = (SeekBar) view;
        switch (seekBar.getId()) {

            case R.id.seekBarHp :
                tvHp.setText(seekBar.getProgress());
                break;

            case R.id.seekBarPrestige :
                tvPrestige.setText(seekBar.getProgress());
                break;

            case R.id.seekBarIntelligence :
                tvIntelligence.setText(seekBar.getProgress());
                break;

            case R.id.seekBarSupport :
                tvSupport.setText(seekBar.getProgress());
                break;

            case R.id.seekBarStrength :
                tvStrength.setText(seekBar.getProgress());
                break;
        }
    }


}

File: app/src/main/java/com/summer/itis/summerproject/ui/cards/add_card/AddCardActivity.kt
package com.summer.itis.summerproject.ui.cards.add_card

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.widget.Toolbar
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.SeekBar
import android.widget.TextView

import com.arellomobile.mvp.presenter.InjectPresenter
import com.summer.itis.summerproject.R
import com.summer.itis.summerproject.model.Card
import com.summer.itis.summerproject.model.pojo.opensearch.Item
import com.summer.itis.summerproject.model.pojo.query.Page
import com.summer.itis.summerproject.ui.base.BaseActivity
import com.summer.itis.summerproject.ui.cards.add_card_list.AddCardListActivity.Companion.CARD_EXTRA
import com.summer.itis.summerproject.ui.tests.add_test.AddTestActivity
import com.summer.itis.summerproject.utils.Const
import com.summer.itis.summerproject.utils.Const.TAG_LOG

import com.summer.itis.summerproject.utils.Const.gsonConverter
import kotlinx.android.synthetic.main.layout_add_card.*

class AddCardActivity : BaseActivity(), AddCardView, SeekBar.OnSeekBarChangeListener {

    private var card: Card? = null

    private var toolbar: Toolbar? = null

    private lateinit var seekBars: List<SeekBar>
    private var balance: Int = 50

    @InjectPresenter
    lateinit var presenter: AddCardPresenter

    private var item: Item? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_add_card)

        presenter = AddCardPresenter(this)
        card = Card()
        seekBars = listOf<SeekBar>(seekBarSupport,seekBarIntelligence,seekBarPrestige,seekBarHp,seekBarStrength)
        item = gsonConverter.fromJson(intent.getStringExtra(ITEM_JSON), Item::class.java)
        card?.abstractCard?.wikiUrl = item!!.url!!.content
        card?.abstractCard?.description = item!!.description!!.content
        item!!.text!!.content?.let { presenter.query(it) }
        initViews()
        setListeners()
    }

    private fun initViews() {
        findViews()
        //        supportActionBar(toolbar);
        setSupportActionBar(toolbar)
        setBackArrow(toolbar!!)

    }

    private fun setListeners() {
        seekBarStrength!!.setOnSeekBarChangeListener(this)
        seekBarHp!!.setOnSeekBarChangeListener(this)
        seekBarPrestige!!.setOnSeekBarChangeListener(this)
        seekBarIntelligence!!.setOnSeekBarChangeListener(this)
        seekBarSupport!!.setOnSeekBarChangeListener(this)

    }

    override fun onStartTrackingTouch(seekBar: SeekBar?) {

    }

    override fun onStopTrackingTouch(seekBar: SeekBar?) {

    }

    override fun onProgressChanged(seekBar: SeekBar?, progress: Int, fromUser: Boolean) {
        changeTvSeekbar(seekBar)
    }

    private fun changeTvSeekbar(seekBar: SeekBar?) {
        val strProgress: String = seekBar?.progress.toString()
        when (seekBar?.id) {
            R.id.seekBarHp -> {
                tvHp!!.setText(strProgress)
                balanceWithOthers(seekBar)
            }

            R.id.seekBarPrestige -> {
                tvPrestige!!.setText(strProgress)
            }

            R.id.seekBarIntelligence -> {
                tvIntelligence!!.setText(strProgress)
            }

            R.id.seekBarSupport -> {
                tvSupport!!.setText(strProgress)
            }

            R.id.seekBarStrength -> {
                tvStrength!!.setText(strProgress)
            }
        }
    }

    private fun balanceWithOthers(seekBar: SeekBar?) {
        while(balance != 50) {
            for (seek in seekBars) {
                if(balance == 50) {
                    return
                }
                if (!seek.equals(seekBar) && balance !=50) {
                    seek.progress--
                }
            }
        }
    }

    override fun setQueryResults(list: List<Page>) {
        val page = list[0]
        card!!.abstractCard?.name = page.title
        card!!.abstractCard?.photoUrl = page.original!!.source
        card!!.abstractCard?.extract = page.extract!!.content

    }

    override fun handleError(throwable: Throwable) {

    }

    private fun findViews() {
        toolbar = findViewById(R.id.toolbar)

    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.edit_menu, menu)

        val checkItem = menu.findItem(R.id.action_check)
        checkItem.setOnMenuItemClickListener {
            prepareCard()
            val intent = Intent()
            Log.d(TAG_LOG,"wiki url = " + card?.abstractCard?.wikiUrl)
            val cardJson = gsonConverter.toJson(card)
            intent.putExtra(CARD_EXTRA, cardJson)
            setResult(Activity.RESULT_OK, intent)
            finish()

            true
        }

        return super.onCreateOptionsMenu(menu)
    }

    private fun prepareCard() {
        card?.hp = seekBarHp.progress
        card?.strength = seekBarStrength.progress
        card?.support = seekBarSupport.progress
        card?.prestige = seekBarPrestige.progress
        card?.intelligence = seekBarIntelligence.progress
    }

    /* override fun onClick(view: View) {

         }
     }*/

    companion object {

        val CARD_EXTRA = "card"

        var ITEM_JSON = "item_json"

        fun start(activity: Activity, item: Item) {
            val intent = Intent(activity, AddTestActivity::class.java)
            val itemJson = gsonConverter.toJson(item)
            intent.putExtra(ITEM_JSON, itemJson)
            activity.startActivity(intent)
        }
    }


}
